### PR TITLE
fix(augmentation): make RandomJigsaw ensure_perm reject the image-preserving permutation

### LINKED ISSUE #4703: RandomJigsaw: ensure_perm=True does not exclude the image-preserving permutation (it rejects arange(n), which transposes the patch grid)
## Summary

`RandomJigsaw(ensure_perm=True)` (the default) is documented as ensuring "the nonidentical patch permutation generation against the original one", i.e. that the output is never the input. It does not: for any grid with both sides greater than 1 the image comes back **unchanged** at the same rate with `ensure_perm=True` as with `ensure_perm=False`.

```python
import torch, kornia.augmentation as K
x = torch.arange(4.).view(2, 2).repeat_interleave(2, 0).repeat_interleave(2, 1)[None, None]   # 2x2 patches of 2x2 px
for ep in (True, False):
    aug = K.RandomJigsaw(grid=(2, 2), p=1.0, ensure_perm=ep); same = 0
    for s in range(2000):
        torch.manual_seed(s); same += int(torch.equal(aug(x), x))
    print(ep, same)
# True  88     <- 4.4 % of calls are a no-op, always with permutation [0, 2, 1, 3]
# False 86
```

Expected rate for a uniform permutation of 4 patches is `1/24 = 4.2 %`; with one permutation excluded, `1/23 = 4.3 %`. So `ensure_perm` excludes a permutation, just not the one that reproduces the image. On a `(2, 3)` grid the no-op rate is 3/2000 either way; only `1 x N` / `N x 1` grids behave as documented (0/2000 vs 1012/2000 for `(1, 2)`).

Measured on `main` @ ff204fb9c, torch 2.14.0, CPU float32.

## Cause

`kornia/augmentation/random_generator/utils.py::randperm` rejects `perm == arange(n)`. `RandomJigsaw.apply_transform` does not use `arange(n)` as the identity, though: an entry's **position** picks the destination cell in column-major order while its **value** indexes the source patch in row-major order (`unfold` yields row-major patches; the final `reshape(-1, b, grid[1], h, piece_w).permute(...)` lays them out column by column). The two orders differ, so

```python
aug = K.RandomJigsaw(grid=(2, 2), p=1.0)
ones = torch.ones(1)
aug(x, params={"batch_prob": ones, "permutation": torch.tensor([[0, 1, 2, 3]])})[0, 0, ::2, ::2]   # [[0, 2], [1, 3]]  transposed
aug(x, params={"batch_prob": ones, "permutation": torch.tensor([[0, 2, 1, 3]])})[0, 0, ::2, ::2]   # [[0, 1], [2, 3]]  the input
```

The image-preserving permutation is `arange(n).view(rows, cols).T.flatten()`, which `ensure_perm` never excludes. The index-identity it does exclude is a genuine shuffle (a transpose of the patch grid on a square grid).

## Expected

Either make the layout consistent, so that `arange(n)` is the identity (and `ensure_perm` is then correct as written), or have `ensure_perm` reject `arange(n).view(rows, cols).T.flatten()`. The first also removes the surprising "identity permutation transposes the image" behaviour; it changes which output a recorded `permutation` replays, so it needs a changelog entry.

Found while reviewing #4634 (batch 6d of #4407), which documents the position/value layout and now cites this issue.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by RizgarOzan**
## What does this pull request do?

`RandomJigsaw(ensure_perm=True)` is meant to never return a selected sample unchanged, but it rejected `arange(N)`. Because a permutation entry's position picks the destination cell column-major while its value indexes the source patch row-major (the layout described in the Convention block), `arange(N)` is the no-op only for a single-row or single-column grid. On a `2 x 2` grid it transposes the patch grid, and the real no-op `[0, 2, 1, 3]` was still drawn.

`JigsawGenerator` now passes the image-preserving permutation, `arange(N).view(rows, columns).T.flatten()`, to `randperm`, which rejects that instead of `arange(N)` (the default is unchanged). This is the second option in the issue. It leaves the layout alone, so recorded `permutation` values replay exactly as before. Changelog fragment: `changelog.d/4708.fixed.md`.

Tests:
- `test_wart_jigsaw_ensure_perm_rejects_the_wrong_permutation_4703` is inverted into `test_convention_jigsaw_ensure_perm_rejects_the_image_preserving_permutation_4703`. On a `2 x 2` grid `[0, 2, 1, 3]` is never drawn and the other 23 permutations all are. `(1, 2)` and `(2, 1)` grids only draw the swap, and a batch of 64 never comes back unchanged.
- `TestRandomJigsaw::test_numerical` had the bug in its golden. With seed 22 the second sample drew `[0, 2, 1, 3]`, and the expected tensor was that sample's input, unchanged. It now draws `[2, 0, 1, 3]`, and I checked the new expected values by hand against the Convention layout.

## Related issue or discussion

Fixes #4703

## What did you check?

Issue reproduction, 2000 seeds, count of calls that return the input unchanged:

| grid | `ensure_perm` | main | this PR |
| --- | --- | --- | --- |
| (2, 2) | True | 88 | **0** |
| (2, 2) | False | 86 | 86 |
| (2, 3) | True | 3 | **0** |
| (1, 2) | True | 0 | 0 |
| (1, 2) | False | 1012 | 1012 |

```text
$ python -m pytest tests/augmentation/test_conventions_mix.py -k 4703      # main: 1 failed; this PR: 1 passed
$ python -m pytest tests/augmentation/test_augmentation_mix.py tests/augmentation/test_conventions_mix.py tests/augmentation/test_base.py \
    --dtype=float32,float64,float16,bfloat16 -k "Jigsaw or jigsaw or 4703 or rng or RNG or fingerprint"
192 passed
$ python -m pytest tests/augmentation
4 failed, 3831 passed, 3403 skipped, 31 xfailed, 5 xpassed
$ python -m pytest --doctest-modules kornia/augmentation/_2d/mix/jigsaw.py kornia/augmentation/random_generator/   # 4 passed
$ python -m pytest tests/test_api_surface.py                                                                     # 148 passed
$ ty check kornia                                                                                                # All checks passed!
$ pre-commit run --files <changed files>                                                                         # all hooks pass
```

The 4 failures are the `*_cannot_be_pickled_4469` wart tests. They fail the same way on `main` in my environment (Windows, Python 3.14, torch 2.14.0 CPU), so they are not related to this change. The RNG fingerprint tests pass: the number of `randperm` draws is unchanged except when the rejected permutation comes up.

Not changed here: a `grid=(1, 1)` with the default `ensure_perm=True` hangs on `main`, because the only permutation is the one being rejected. It hangs the same way after this change. Should that raise a `ValueError` in the constructor? I can do it in a follow-up if you want.

## How did AI help?

I made this with Claude Code (an AI coding agent). It reproduced the issue, wrote the change and the tests, and ran every check listed above. The numbers in this description come from those runs.

### CODE REVIEW

**Review by ducha-aiki — DISMISSED** [review]
The revised identity permutation matches the Jigsaw transform layout for rectangular as well as square grids, so `ensure_perm` excludes the image-preserving arrangement. No introduced blocking code finding identified. The existing `grid=(1,1), ensure_perm=True` nontermination limitation predates this patch; the two unresolved Copilot conversations about it remain a conversation gate before merge.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6-astra).

**Comment by ducha-aiki** [comment]
## Review of #4708 (Copilot findings checked, plus my own review) at `da4f20f8e`

**Verdict: approve.** The fix is correct. I checked both Copilot findings: they are real, but they are not regressions from this PR. I pushed a guard that resolves them (`da4f20f8e`, details below).

### The fix is correct, including on rectangular grids
I ran every permutation through `RandomJigsaw.apply_transform` on an image with a distinct value in each cell, to find the true no-op. Then I compared it with the permutation the generator now rejects, and counted unchanged outputs in 2000 draws (`ensure_perm=True`, `p=1`). The script ran from each tree's root, and `kornia.__file__` confirmed which tree was imported:

| grid | true no-op | `arange(N).view(grid).T.flatten()` | unchanged, main | unchanged, PR |
|---|---|---|---|---|
| (2, 2) | (0, 2, 1, 3) | (0, 2, 1, 3) | 79 | **0** |
| (2, 3) | (0, 3, 1, 4, 2, 5) | same | 3 | **0** |
| (3, 2) | (0, 2, 4, 1, 3, 5) | same | 3 | **0** |
| (1, 3) / (3, 1) | (0, 1, 2) | same | 0 | 0 |

`randperm`'s default (`identity=None` → `arange(n)`) is unchanged. `JigsawGenerator` is the only caller of the helper; `color_jiggle`'s `self.randperm` is a different function. `identity` is built on `self._device`, the same device `torch.randperm` draws on, so `torch.equal` compares like with like.

### The pins can fail
I reverted `identity` to `arange(perm_times)` and kept the new tests. Both failed: `test_convention_jigsaw_ensure_perm_rejects_the_image_preserving_permutation_4703` (`(0, 2, 1, 3)` drawn) and `TestRandomJigsaw::test_numerical`. With the fix restored, the Jigsaw, jigsaw and 4703 tests in `test_augmentation_mix.py` and `test_conventions_mix.py` pass: 187 passed on CPU in float32, float64, float16 and bfloat16. `test_base.py` and `test_api_surface.py` also pass (248 passed). `grep -rnE "#4703|issues/4703" kornia/ tests/ docs/` is empty after the PR.

### Copilot findings (both about `grid=(1, 1)` hanging): confirmed, predate this PR, now fixed
`RandomJigsaw(grid=(1, 1), p=1.0)(torch.rand(1, 1, 4, 4))` never returns on `main` or on `6eec922d`: a 20 s alarm killed it on both. The only permutation of one patch is the one `ensure_perm` rejects. So this PR did not introduce the hang. But the new docstring promises that "a selected sample is never returned unchanged", and a 1 x 1 grid cannot keep that promise. The author already offered the `ValueError` as a follow-up. It is three lines, so I pushed it here instead:

- `JigsawGenerator.__init__` raises `ValueError("ensure_perm=True needs at least two patches, ...")` when `grid[0] * grid[1] == 1` and `ensure_perm` is set. `ensure_perm=False` still works on a 1 x 1 grid and returns the input.
- The `ensure_perm` docstring and `changelog.d/4708.fixed.md` state the constraint.
- `TestRandomJigsaw::test_single_cell_grid` checks both halves. Its `pytest.raises` is on the constructor, so without the guard it fails at once instead of hanging. I checked that: it fails at `6eec922d` and passes at `da4f20f8e`. Pre-commit passes on the changed files.

Copilot's other point, that `identity` is allocated even when `ensure_perm=False`: I am dropping it. It is one `arange(N)` per forward call, next to N `randperm` draws. Hoisting it would add a branch for no measurable gain.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

File: kornia/augmentation/_2d/mix/jigsaw.py [hunk 1]
```diff
@@ -39,9 +39,9 @@ class RandomJigsaw(MixAugmentationBaseV2):
     Args:
         grid: the Jigsaw puzzle grid. e.g. (2, 2) means
             each output will mix image patches in a 2x2 grid.
-        ensure_perm: reject the index-identity permutation ``[0, ..., N - 1]`` when drawing. That is the
-            image-preserving permutation only for a single-row or single-column grid; for any other grid the
-            output can still equal the input (`#4703 <https://github.com/kornia/kornia/issues/4703>`_).
+        ensure_perm: reject the image-preserving permutation ``arange(N).view(rows, columns).T.flatten()`` when
+            drawing, so a selected sample is never returned unchanged. For a single-row or single-column grid that
+            permutation is ``[0, ..., N - 1]``; see the Convention block.
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/augmentation/_2d/mix/jigsaw.py]: Confirmed, but not introduced by this PR: the call hangs on `main` too, because the only permutation of one patch is the rejected one. Fixed in da4f20f8e. `JigsawGenerator.__init__` now raises `ValueError` for a 1 x 1 grid with `ensure_perm=True`; the docstring and changelog say so, and `TestRandomJigsaw::test_single_cell_grid` pins it. `ensure_perm=False` on a 1 x 1 grid still returns the input.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

File: kornia/augmentation/random_generator/_2d/jigsaw.py [hunk 2]
```diff
@@ -64,14 +64,20 @@ def forward(self, batch_shape: Tuple[int, ...], same_on_batch: bool = False) ->
...
         perm_times = self.grid[0] * self.grid[1]
+        # RandomJigsaw fills destination cells column-major but indexes source patches row-major, so the
+        # image-preserving permutation is the transposed index grid, not arange(perm_times).
+        identity = torch.arange(perm_times, device=self._device).view(self.grid).T.flatten()
         # Generate mosiac order in one shot
         if batch_size == 0:
             rand_ids = torch.zeros([0, perm_times], device=self._device)
         elif same_on_batch:
-            rand_ids = randperm(perm_times, ensure_perm=self.ensure_perm, device=self._device)
+            rand_ids = randperm(perm_times, ensure_perm=self.ensure_perm, identity=identity, device=self._device)
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/augmentation/random_generator/_2d/jigsaw.py:75]: Confirmed, but not introduced by this PR: the call hangs on `main` too, because the only permutation of one patch is the rejected one. Fixed in da4f20f8e. `JigsawGenerator.__init__` now raises `ValueError` for a 1 x 1 grid with `ensure_perm=True`; the docstring and changelog say so, and `TestRandomJigsaw::test_single_cell_grid` pins it. `ensure_perm=False` on a 1 x 1 grid still returns the input.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Review by ducha-aiki — APPROVED** [review]
Recording the explicit approval from the linked completed review at this head as a formal GitHub review. That review covers rectangular-grid correctness, regression validation, and the added single-cell guard resolving the inherited nontermination case.

[Completed review](https://github.com/kornia/kornia/pull/4708#issuecomment-5760921624).

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6-astra).

**Comment by ducha-aiki** [comment]
Merged current `main` (088586663, #4711) into this branch as c2230d6a6. The previous red CI was not caused by this PR: `main` itself was failing on four stale batch-6d convention pins (#4655 / #4656 / #4649 and the RandAugment formula pin), and #4711 fixed them. The merge was clean, and `git diff main HEAD` is exactly this PR's own diff.

Checked locally on the merge tree (cpu, float32/float64): `tests/augmentation/test_augmentation_mix.py`, `test_conventions_mix.py` and `test_conventions_auto.py` all pass (443 passed). Ruff check and format are clean on the touched files.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
